pkg/cloud/cloud.go

```diff
@@ -549,3 +680,56 @@ func isAWSErrorVolumeNotFound(err error) bool {
 	}
 	return false
 }
+
+// Helper function for describeSnapshot callers. Tries to retype given error to AWS error
+// and returns true in case the AWS error is "InvalidSnapshot.NotFound", false otherwise
+func isAWSErrorSnapshotNotFound(err error) bool {
+	if awsError, ok := err.(awserr.Error); ok {
+		// https://docs.aws.amazon.com/AWSEC2/latest/APIReference/errors-overview.html
+		if awsError.Code() == "InvalidSnapshot.NotFound" {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (c *cloud) waitForSnapshotCreate(ctx context.Context, snapshotID *string) error {
+	// This should give about 1 minute maximal interval
+	backoff := wait.Backoff{
+		Duration: 1 * time.Second,
+		Factor:   1.5,
+		Steps:    10,
+	}
+	request := &ec2.DescribeSnapshotsInput{
+		SnapshotIds: []*string{
+			snapshotID,
+		},
+	}
+
+	conditionFunc := func() (done bool, err error) {
+		snapshot, err := c.getSnapshot(ctx, request)
+		if err != nil {
+			return true, err
+		}
+		if snapshot.State != nil {
+			switch *snapshot.State {
+			case "completed":
+				return true, nil
+			case "pending":
+				return false, nil
+			default:
+				return true, fmt.Errorf("unexpected State of newly created AWS EBS snapshot %v: %q", snapshotID, *snapshot.State)
+			}
+		}
+		return false, nil
+	}
+
+	// Truncated exponential backoff: if the exponential backoff times-out, just keep polling using the longest interval
+	err := wait.ExponentialBackoff(backoff, conditionFunc)
+	if err == wait.ErrWaitTimeout {
+		timeout := time.Duration(backoff.Duration.Seconds() * math.Pow(backoff.Factor, float64(backoff.Steps)))
```

Comment:
Kuberentes API is supposed to be "intent based" so as long as there exists a VolumeSnapshot object the API should try to create it (no timeout). The idea here is not to extend the wait period too much so in case the operation succeeds/fail we are able to inform the user. Since it's unknown how long may the operation take I only used the exponential backoff for the first few poll iterations and then stopped to prolong the interval and make it constant.